hazelcast/src/main/java/com/hazelcast/spi/impl/operationexecutor/classic/ClassicOperationExecutor.java

```diff
@@ -248,18 +248,25 @@ public boolean isInvocationAllowedFromCurrentThread(Operation op, boolean isAsyn
             return true;
         }
 
+        // allowed to invoke non partition specific task
         if (op.getPartitionId() < 0) {
             return true;
         }
 
-        // we are allowed to invoke from non PartitionOperationThreads (including GenericOperationThread)
+        // allowed to invoke from non PartitionOperationThreads (including GenericOperationThread)
         if (!(currentThread instanceof PartitionOperationThread)) {
             return true;
         }
 
-        // we are only allowed to invoke from a PartitionOperationThread if the operation belongs to that
-        // PartitionOperationThread.
+        // if already running an operation for a partition (id >= 0) it means that it's a nested invocation
+        // in this case partitionId of both nested operations has to match
         PartitionOperationThread partitionThread = (PartitionOperationThread) currentThread;
+        OperationRunner runner = partitionThread.getCurrentOperationRunner();
+        if (runner != null) {
+            return runner.getPartitionId() == op.getPartitionId();
+        }
+
+        // TODO: this is racy... what if the partition is being migrated at the same time?
         return toPartitionThreadIndex(op.getPartitionId()) == partitionThread.threadId;
```

Comment:
I think we should. We should also update the documentation to clearly explain what this method is doing.